Change MaybeAdd to use a non-blocking send for the incoming channel

kkaneda opened the pull request:
When the buffer size is 10, the test added by this PR has the following failure:

1) `maybeSplitRangesByConfigs()` acquires the mutex lock.
2) `maybeSplitRangesByConfigs()` blocks since the channel is full.
3) the split queue won't finish processing the range since the range split never finishes. 

This queue processing never finishes probably because `GroupStorage` gets stuck with the deadlock. Here are stacktraces of goruntines from the test failure:

```
goroutine 245 [chan send]:
github.com/cockroachdb/cockroach/storage.(*baseQueue).MaybeAdd(0xc20805e900, 0xc208186510, 0x0, 0x809, 0x0, 0x0, 0x0)
     /Users/kaneda/Development/go/src/github.com/cockroachdb/cockroach/storage/queue.go:177 +0x4b0
github.com/cockroachdb/cockroach/storage.(*Store).maybeSplitRangesByConfigs(0xc2080a4000, 0xc208070a00, 0x1f, 0x20)
     /Users/kaneda/Development/go/src/github.com/cockroachdb/cockroach/storage/store.go:679 +0x284
github.com/cockroachdb/cockroach/storage.(*Store).configGossipUpdate(0xc2080a4000, 0x4abb2d0, 0xa, 0xc208022301)
     /Users/kaneda/Development/go/src/github.com/cockroachdb/cockroach/storage/store.go:640 +0x302
```

```
goroutine 20 [semacquire]:
sync.(*Mutex).Lock(0xc2080a4138)
     /usr/local/go/src/sync/mutex.go:66 +0xd3
sync.(*RWMutex).Lock(0xc2080a4138)
     /usr/local/go/src/sync/rwmutex.go:82 +0x33
github.com/cockroachdb/cockroach/storage.(*Store).GroupStorage(0xc2080a4000, 0x1, 0x0, 0x0)
     /Users/kaneda/Development/go/src/github.com/cockroachdb/cockroach/storage/store.go:1437 +0x56
github.com/cockroachdb/cockroach/multiraft.func·008()
     /Users/kaneda/Development/go/src/github.com/cockroachdb/cockroach/multiraft/storage.go:157 +0x3ac
```

```
goroutine 23 [chan send]:
github.com/cockroachdb/cockroach/multiraft.(*MultiRaft).SubmitCommand(0xc20809e180, 0x19, 0xc20815e4b0, 0x10, 0xc20827cc40, 0xdc, 0xdc, 0xc20811dc80)
    /Users/kaneda/Development/go/src/github.com/cockroachdb/cockroach/multiraft/multiraft.go:333 +0x3d0
   ...
github.com/cockroachdb/cockroach/client.runOne(0x5715520, 0xc2080400c8, 0xc208244c00, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
    /Users/kaneda/Development/go/src/github.com/cockroachdb/cockroach/client/db.go:900 +0x6b
github.com/cockroachdb/cockroach/client.(*DB).AdminSplit(0xc2080400c8, 0x4a369c0, 0xc208285360, 0x4a369c0, 0xc208285380, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
    /Users/kaneda/Development/go/src/github.com/cockroachdb/cockroach/client/db.go:287 +0xfb
github.com/cockroachdb/cockroach/storage.(*splitQueue).process(0xc208020d60, 0x0, 0x716, 0x0, 0x0, 0x0, 0xc208130e10, 0x0, 0x0)
    /Users/kaneda/Development/go/src/github.com/cockroachdb/cockroach/storage/split_queue.go:92 +0x32c
github.com/cockroachdb/cockroach/storage.(*baseQueue).processOne(0xc20805e900, 0xc208076000, 0xc20805e480)
    /Users/kaneda/Development/go/src/github.com/cockroachdb/cockroach/storage/queue.go:261 +0x4dc
github.com/cockroachdb/cockroach/storage.func·013()
    /Users/kaneda/Development/go/src/github.com/cockroachdb/cockroach/storage/queue.go:214 +0x14b
```

Increased the channel size to 200 since `splitQueueMaxSize` is 100, but I'm not sure what's the right fix. Maybe we should try releasing the lock from `maybeSplitRangesByConfigs()`.

kkaneda committed:
Add a test to reproduce a deadlock when the queue size is small

When the buffer size of the incoming channel is 10, the test fails
with the following deadlock issue:

1) maybeSplitRangesByConfigs() acquires the mutex lock.
2) maybeSplitRangesByConfigs() blocks since the channel is full.
3) the split queue won't finish processing the range since the range
   split never finishes.

Here are stacktraces of goruntines from the test failure:

goroutine 245 [chan send]:
github.com/cockroachdb/cockroach/storage.(*baseQueue).MaybeAdd(0xc20805e900, 0xc208186510, 0x0, 0x809, 0x0, 0x0, 0x0)
     /Users/kaneda/Development/go/src/github.com/cockroachdb/cockroach/storage/queue.go:177 +0x4b0
github.com/cockroachdb/cockroach/storage.(*Store).maybeSplitRangesByConfigs(0xc2080a4000, 0xc208070a00, 0x1f, 0x20)
     /Users/kaneda/Development/go/src/github.com/cockroachdb/cockroach/storage/store.go:679 +0x284
github.com/cockroachdb/cockroach/storage.(*Store).configGossipUpdate(0xc2080a4000, 0x4abb2d0, 0xa, 0xc208022301)
     /Users/kaneda/Development/go/src/github.com/cockroachdb/cockroach/storage/store.go:640 +0x302

goroutine 20 [semacquire]:
sync.(*Mutex).Lock(0xc2080a4138)
     /usr/local/go/src/sync/mutex.go:66 +0xd3
sync.(*RWMutex).Lock(0xc2080a4138)
     /usr/local/go/src/sync/rwmutex.go:82 +0x33
github.com/cockroachdb/cockroach/storage.(*Store).GroupStorage(0xc2080a4000, 0x1, 0x0, 0x0)
     /Users/kaneda/Development/go/src/github.com/cockroachdb/cockroach/storage/store.go:1437 +0x56
github.com/cockroachdb/cockroach/multiraft.func·008()
     /Users/kaneda/Development/go/src/github.com/cockroachdb/cockroach/multiraft/storage.go:157 +0x3ac

goroutine 23 [chan send]:
github.com/cockroachdb/cockroach/multiraft.(*MultiRaft).SubmitCommand(0xc20809e180, 0x19, 0xc20815e4b0, 0x10, 0xc20827cc40, 0xdc, 0xdc, 0xc20811dc80)
    /Users/kaneda/Development/go/src/github.com/cockroachdb/cockroach/multiraft/multiraft.go:333 +0x3d0
   ...
github.com/cockroachdb/cockroach/client.runOne(0x5715520, 0xc2080400c8, 0xc208244c00, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
    /Users/kaneda/Development/go/src/github.com/cockroachdb/cockroach/client/db.go:900 +0x6b
github.com/cockroachdb/cockroach/client.(*DB).AdminSplit(0xc2080400c8, 0x4a369c0, 0xc208285360, 0x4a369c0, 0xc208285380, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
    /Users/kaneda/Development/go/src/github.com/cockroachdb/cockroach/client/db.go:287 +0xfb
github.com/cockroachdb/cockroach/storage.(*splitQueue).process(0xc208020d60, 0x0, 0x716, 0x0, 0x0, 0x0, 0xc208130e10, 0x0, 0x0)
    /Users/kaneda/Development/go/src/github.com/cockroachdb/cockroach/storage/split_queue.go:92 +0x32c
github.com/cockroachdb/cockroach/storage.(*baseQueue).processOne(0xc20805e900, 0xc208076000, 0xc20805e480)
    /Users/kaneda/Development/go/src/github.com/cockroachdb/cockroach/storage/queue.go:261 +0x4dc
github.com/cockroachdb/cockroach/storage.func·013()
    /Users/kaneda/Development/go/src/github.com/cockroachdb/cockroach/storage/queue.go:214 +0x14b

vivekmenezes commented:
Does the range have to be added to the queue to signal processing?

If not then we can change the send to use select.

https://gobyexample.com/non-blocking-channel-operations

But yeah we should not hold on to a lock when the code can block.

If adding the range to the channel is required, then perhaps the function
should return with error, the caller unlocks on receiving an error and then
retries again in a bit.

Vivek

On Wed, Jun 3, 2015 at 3:38 AM kkaneda <notifications@github.com> wrote:

> When the buffer size is 10, the test added by this PR has the following
> failure:
>
> 1) maybeSplitRangesByConfigs() acquires the mutex lock.
> 2) maybeSplitRangesByConfigs() blocks since the channel is full.
> 3) the split queue won't finish processing the range since the range split
> never finishes.
>
> This queue processing never finishes probably because GroupStorage gets
> stuck with the deadlock. Here are stacktraces of goruntines from the test
> failure:
>
> goroutine 245 [chan send]:github.com/cockroachdb/cockroach/storage.(*baseQueue).MaybeAdd(0xc20805e900, 0xc208186510, 0x0, 0x809, 0x0, 0x0, 0x0)
>      /Users/kaneda/Development/go/src/github.com/cockroachdb/cockroach/storage/queue.go:177 +0x4b0github.com/cockroachdb/cockroach/storage.(*Store).maybeSplitRangesByConfigs(0xc2080a4000, 0xc208070a00, 0x1f, 0x20)
>      /Users/kaneda/Development/go/src/github.com/cockroachdb/cockroach/storage/store.go:679 +0x284github.com/cockroachdb/cockroach/storage.(*Store).configGossipUpdate(0xc2080a4000, 0x4abb2d0, 0xa, 0xc208022301)
>      /Users/kaneda/Development/go/src/github.com/cockroachdb/cockroach/storage/store.go:640 +0x302
>
> goroutine 20 [semacquire]:
> sync.(*Mutex).Lock(0xc2080a4138)
>      /usr/local/go/src/sync/mutex.go:66 +0xd3
> sync.(*RWMutex).Lock(0xc2080a4138)
>      /usr/local/go/src/sync/rwmutex.go:82 +0x33github.com/cockroachdb/cockroach/storage.(*Store).GroupStorage(0xc2080a4000, 0x1, 0x0, 0x0)
>      /Users/kaneda/Development/go/src/github.com/cockroachdb/cockroach/storage/store.go:1437 +0x56github.com/cockroachdb/cockroach/multiraft.func·008() <http://github.com/cockroachdb/cockroach/multiraft.func%C2%B7008()>
>      /Users/kaneda/Development/go/src/github.com/cockroachdb/cockroach/multiraft/storage.go:157 +0x3ac
>
> goroutine 23 [chan send]:github.com/cockroachdb/cockroach/multiraft.(*MultiRaft).SubmitCommand(0xc20809e180, 0x19, 0xc20815e4b0, 0x10, 0xc20827cc40, 0xdc, 0xdc, 0xc20811dc80)
>     /Users/kaneda/Development/go/src/github.com/cockroachdb/cockroach/multiraft/multiraft.go:333 +0x3d0
>    ...github.com/cockroachdb/cockroach/client.runOne(0x5715520, 0xc2080400c8, 0xc208244c00, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
>     /Users/kaneda/Development/go/src/github.com/cockroachdb/cockroach/client/db.go:900 +0x6bgithub.com/cockroachdb/cockroach/client.(*DB).AdminSplit(0xc2080400c8, 0x4a369c0, 0xc208285360, 0x4a369c0, 0xc208285380, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
>     /Users/kaneda/Development/go/src/github.com/cockroachdb/cockroach/client/db.go:287 +0xfbgithub.com/cockroachdb/cockroach/storage.(*splitQueue).process(0xc208020d60, 0x0, 0x716, 0x0, 0x0, 0x0, 0xc208130e10, 0x0, 0x0)
>     /Users/kaneda/Development/go/src/github.com/cockroachdb/cockroach/storage/split_queue.go:92 +0x32cgithub.com/cockroachdb/cockroach/storage.(*baseQueue).processOne(0xc20805e900, 0xc208076000, 0xc20805e480)
>     /Users/kaneda/Development/go/src/github.com/cockroachdb/cockroach/storage/queue.go:261 +0x4dcgithub.com/cockroachdb/cockroach/storage.func·013() <http://github.com/cockroachdb/cockroach/storage.func%C2%B7013()>
>     /Users/kaneda/Development/go/src/github.com/cockroachdb/cockroach/storage/queue.go:214 +0x14b
>
> Increased the channel size to 200 since splitQueueMaxSize is 100, but I'm
> not sure what's the right fix. Maybe we should try releasing the lock from
> maybeSplitRangesByConfigs().
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/cockroachdb/cockroach/pull/1240
> Commit Summary
>
>    - Increase the buffer size of the incoming channel of the queue
>
> File Changes
>
>    - *M* storage/client_split_test.go
>    <https://github.com/cockroachdb/cockroach/pull/1240/files#diff-0> (56)
>    - *M* storage/queue.go
>    <https://github.com/cockroachdb/cockroach/pull/1240/files#diff-1> (2)
>
> Patch Links:
>
>    - https://github.com/cockroachdb/cockroach/pull/1240.patch
>    - https://github.com/cockroachdb/cockroach/pull/1240.diff
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/1240>.
>

kkaneda commented:
Ah, non-blocking example looks nice!

My understanding was that we don't necessarily add ranges to the queue. (We anyway might drop ranges if the queue size reaches `maxSize`). If that's correct, will make the following change:

- Change `MaybeAdd` to return an error when the channel is full:
- Change `maybeSplitRangesByConfigs` to just ignore errors or retry on errors.

vivekmenezes commented:
One solution is to move grabbing the lock within

for _, config := range configMap {

for processing each entry in the config.

And everytime s.splitQueue().MaybeAdd() returns an error, unlock, timeout
and retry applying the config change.

But you also have to worry about two config-map changes being applied
interleaved, and that might not be good.

It is best we understand the requirements here before we make this change.

1. Can we push back a configmap change to the network in case of failure,
or do you need to retry applying the configmap change piecemeal?

2. Can a configmap change be applied piecemeal in case of failure, or do
you need to hold on to the lock during the entire change application?

Vivek

On Wed, Jun 3, 2015 at 10:18 AM kkaneda <notifications@github.com> wrote:

> Ah, non-blocking example looks nice!
>
> My understanding was that we don't necessarily add ranges to the queue.
> (We anyway might drop ranges if the queue size reaches maxSize). If
> that's correct, will make the following change:
>
>    - Change MaybeAdd to return an error when the channel is full:
>    - Change maybeSplitRangesByConfigs to just ignore errors or retry on
>    errors.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/1240#issuecomment-108447219>
> .
>

spencerkimball commented:
Change `baseQueue.MaybeAdd` to use a non-blocking send on the queue's `incoming` channel. Increasing this to 200 doesn't solve the underlying problem of queues filling up and it's also more ranges than we need to have queued. We might increase the queue size to 50 though..

bdarnell commented:
Not all of our queues are the same. rangeGCQueue.shouldQueue nearly always returns true, so we expect to have more ranges in the queue there. If we change to a non-blocking send in MaybeAdd, I'm worried about the possibility that we silently skip ranges in the scanner. It's not a huge problem since the scanner's pace interval should keep things reasonable (and if we drop a range it might get picked up next time), but there's some risk that we systematically miss certain ranges (i.e. the `cap(channel)+1`th range). 

Perhaps MaybeAdd should have a blocking and non-blocking mode. The scanner would use blocking mode (since it doesn't hold any locks that would cause deadlocks and should slow down instead of skipping ranges) while most other callers would use non-blocking mode.

spencerkimball commented:
We should make the `shouldQueue` method of rangeGCQueue more "accurate".. Seems strange to basically queue up every range and its mother when probably none of them is going to be gc'd. This defeats the purpose of `shouldQueue`, and is going to create a real storm of range lookup requests.

spencerkimball commented:
One idea that comes to mind is to look at the expiration of the leader lease. If it's greater than some reasonable threshold (like 10 days, but we'd want to make it variable for tests), then we could consider the range for GC.

bdarnell commented:
Requiring the range to be dormant for some time before queuing seems reasonable, although 10 days seems way too much uncollected garbage in the absence of some other signal that can trigger a likely GC.

kkaneda commented:
Thanks for the inputs! 

@vivekmenezes , I need to look into the code more to answer your question, but my understanding so far is that:

1. It's difficult to push back since the update comes from gossip. 
2. We can applying the configmap change piecemeal. 

We might not want to retry again and again in the gossip update callback as the info might get stale.

Will first try adding a blocking and a non-blocking mode of `MaybeAdd`. Will then take a look at the possible change to the `shouldQueue` method of rangeGCQueue.

Will send a quick PR to increase the queue size. I think there is no downside of doing that..

kkaneda committed:
Change MaybeAdd to use a non-blocking send for the incoming channel

This commit also changes the buffer size of incoming to 1. This is
fine since we don't actually read the content of data received from
the channel. We just need a signal when a new queue is added.

kkaneda commented:
I was totally misunderstanding the code. We don't really use data coming from `incoming`, we just use the channel as a signal for new range addition..

So, I've changed `MaybeAdd` to use a non-blocking send and changed `incoming` to `chan bool` of size `1`. The new test still passed without any deadlock.

I think the issue discussed here is mainly a problem of the actual queue size bounded by `queue.maxSize`. I think it makes sense to make `MaybeAdd` block when it is called from the scanner and when the queue is full (to avoid the skipping problem Ben described).

I'm still not totally sure whether we want to make `MaybeAdd` block in other places (e.g., `maybeSplitRangesByConfigs`), but best-effort (e.g., do not add a range when the queue is full) looks OK to me. The scanner will eventually process every range.

tamird commented:
Probably need to update the comment at https://github.com/cockroachdb/cockroach/blob/kkaneda/many_splits_test/storage/queue.go#L208

tamird commented on storage/client_split_test.go:
```diff
@@ -459,3 +460,54 @@ func TestStoreRangeSplitOnConfigs(t *testing.T) {
+		if err != nil {
+			t.Fatalf("failed to scan meta2 keys: %s", err)
+		}
+		return len(rows) == 24
```
can you make this magic number less magical?

kkaneda replied:
Sure, changed the test to check the expected values of row keys.

tamird commented on storage/client_split_test.go:
```diff
@@ -459,3 +460,54 @@ func TestStoreRangeSplitOnConfigs(t *testing.T) {
+		if err != nil {
+			t.Fatalf("failed to scan meta2 keys: %s", err)
+		}
+		return len(rows) == 64
```
ditto

kkaneda replied:
Done.

kkaneda commented:
Thanks for the review, @tamird . Sure, updated the comment.

```
select {
// Incoming signal sets the next time to process if there were previously
// no ranges in the queue.
}
```

tamird commented on storage/client_split_test.go:
```diff
@@ -459,3 +460,83 @@ func TestStoreRangeSplitOnConfigs(t *testing.T) {
+
+	// Check that we finish splitting in allotted time.
+	expKeys := []proto.Key{}
+	for i := 1; i < 22; i++ {
```
this `22` is still a magic number - perhaps extract `20` above into a variable and then explain why this is 2 more than that value?

kkaneda replied:
Sure, defined a variable for 20 (`numDbs`) and added the following comment:

```
// Expect numDb+1 keys as the zone config for "db20" creates
// "meta2db20" and "meta2db21" as start/end keys.
for i := 1; i <= numDbs+1; i++ {
```

tamird commented on storage/client_split_test.go:
```diff
@@ -459,3 +460,83 @@ func TestStoreRangeSplitOnConfigs(t *testing.T) {
+
+	// Check the result of splits again.
+	expKeys = []proto.Key{}
+	for i := 1; i < 21; i++ {
```
magic number

kkaneda replied:
Changed the for loop to `for i := 1; i <= numDbs; i++`.

bdarnell commented on storage/queue.go:
```diff
@@ -100,7 +100,7 @@ type baseQueue struct {
 	impl       queueImpl
 	maxSize    int                  // Maximum number of ranges to queue
-	incoming   chan *Range          // Channel for ranges to be queued
+	incoming   chan bool            // Channel signalled when a new range is added to the queue.
```
We generally use `chan struct{}` (and `case bq.incoming <- struct{}{}:`) to indicate that the channel just triggers an event and doesn't convey a value (i.e. it doesn't matter whether you send `true` or `false`)

kkaneda replied:
Ah, I see. Fixed.

bdarnell commented:
LGTM

tamird commented:
LGTM
